File: LC-Python-Solution/875. Koko Eating Bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:

        l = 1
        r = max(piles)
        res = r

        def canEat(cap):
            hours = 0
            for pile in piles:
                hours += (pile + cap - 1) // cap  ``
            return hours <= h

        while l<=r:
            cap = l + (r-l)//2
            if canEat(cap):
                res = min(res,cap)
                r = cap - 1
            else:
                l = cap + 1
        
        return res



        




        

File: LC_JS-SOLUTIONS/875. Koko Eating Bananas.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function (piles, h) {
  let l = 1;
  let r = Math.max(...piles);
  let res = r;

  function canEat(cap) {
    let hour = 0;

    for (let ban of piles) {
      hour += Math.floor((ban + cap - 1) / cap);
    }
    return hour <= h;
  }

  while (l <= r) {
    let cap = l + Math.floor((r - l) / 2);

    if (canEat(cap)) {
      res = Math.min(cap, res);
      r = cap - 1;
    } else {
      l = cap + 1;
    }
  }
  return res;
};
